File: src/main/java/com/avs/filmoteca/data/repository/DataRepository.java
package com.avs.filmoteca.data.repository;

import java.util.List;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;

import com.avs.filmoteca.data.domain.Movie;
import com.avs.filmoteca.data.domain.UpdateStatus;
import com.avs.filmoteca.data.domain.mapper.MoviesDataMapper;
import com.avs.filmoteca.data.domain.push.PushMessage;
import com.avs.filmoteca.data.ws.ApiClient;
import com.google.gson.Gson;

import okhttp3.ResponseBody;
import rx.Observable;
import rx.schedulers.Schedulers;

public class DataRepository {

	private static DataRepository sDataRepository;

	private static Preferences sPreferences;

	public static DataRepository getInstance() {
		if (sDataRepository == null)
			sDataRepository = new DataRepository();
		return sDataRepository;
	}

	private DataRepository() {
		sPreferences = Preferences.userNodeForPackage(DataRepository.class);
	}

	public Observable<List<Movie>> getPublishedMoviesObservable() {
		return ApiClient.getFilmotecaInterface().getMoviesListHtmlObservable().map(MoviesDataMapper::transformMovie)
				.subscribeOn(Schedulers.newThread()).observeOn(Schedulers.trampoline());
	}

	public Observable<Void> getUpdateMoviesObservable(List<String> movieTitles) {
		return ApiClient.getBackendInterface().getUpdateMoviesObservable(movieTitles)
				.subscribeOn(Schedulers.newThread()).observeOn(Schedulers.trampoline());
	}

	public Observable<List<String>> getStoredMoviesObservable() {
		return ApiClient.getBackendInterface().getStoredMoviesObservable().subscribeOn(Schedulers.newThread())
				.observeOn(Schedulers.trampoline());
	}

	public Observable<List<String>> getRegistrationIdsObservable() {
		return ApiClient.getBackendInterface().getRegistrationIdsObservable().subscribeOn(Schedulers.newThread())
				.observeOn(Schedulers.trampoline());
	}

	public Observable<ResponseBody> getPushDeliveryObservable(PushMessage message) {
		return ApiClient.getFCMInterface()
				.getPushDeliveryObservable("key=" + System.getenv("FIREBASE_API_KEY"), message)
				.subscribeOn(Schedulers.newThread()).observeOn(Schedulers.trampoline());
	}

	public void setUpdating(boolean updating) {
		sPreferences.putBoolean(UpdateStatus.PREFERENCE_IS_UPDATING, updating);
		try {
			sPreferences.sync();
		} catch (BackingStoreException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}

	public boolean isUpdating() {
		return sPreferences.getBoolean(UpdateStatus.PREFERENCE_IS_UPDATING, false);
	}

}

File: src/main/java/com/avs/filmoteca/data/repository/DataRepository.kt
package com.avs.filmoteca.data.repository

import com.avs.filmoteca.data.Environment
import com.avs.filmoteca.domain.model.Region
import com.avs.filmoteca.domain.model.UpdateStatus
import com.avs.filmoteca.domain.mapper.MoviesDataMapper
import com.avs.filmoteca.domain.model.push.PushMessage
import com.avs.filmoteca.data.ws.ApiClient
import java.util.prefs.BackingStoreException
import java.util.prefs.Preferences

class DataRepository private constructor() {

    companion object {

        private var preferences: Preferences = Preferences.userNodeForPackage(DataRepository::class.java)

        val instance: DataRepository by lazy { DataRepository() }

    }

    fun getPublishedMovies(region: Region): List<String> =
            ApiClient.filmotecaInterface.getMoviesListHtml(region.endpoint).execute().body()?.let { html ->
                MoviesDataMapper.newInstance(region).map(html).map { it.title }
            } ?: emptyList()

    fun getStoredMovies(region: Region): List<String> =
            ApiClient.apiInterface.getStoredMovies(region.code).execute().body() ?: emptyList()

    fun getRegistrationIds(region: Region): List<String> =
            ApiClient.apiInterface.getRegistrationIds(region.code).execute().body() ?: emptyList()

    fun updateMovies(region: Region, movieTitles: List<String>) {
        ApiClient.apiInterface.updateMovies(movieTitles, region.code).execute()
    }

    fun sendPush(registrationIds: List<String>) {
        val groupedRegistrationIds: List<List<String>> =
                registrationIds.groupBy { registrationIds.indexOf(it) / 1000 }.values.toList()
        groupedRegistrationIds.forEach {
            val message = PushMessage.Builder()
                    .setRegistrationIds(it)
                    .setTitleResId("notification_title_normal")
                    .setMessageResId("notification_message_new_movies")
                    .setIconResId("ic_notification").build()
            ApiClient.fcmInterface.sendPush("key=" + Environment.FirebaseApiKey.value, message).execute()
        }
    }

    fun getLastUpdateStatus(region: Region): Int =
            preferences.getInt("${UpdateStatus.PREFERENCE_UPDATE_STATUS}_${region.code}", UpdateStatus.NOT_UPDATING)

    fun setUpdateStatus(region: Region, status: Int) {
        preferences.putInt("${UpdateStatus.PREFERENCE_UPDATE_STATUS}_${region.code}", status)
        try {
            preferences.sync()
        } catch (e: BackingStoreException) {
            e.printStackTrace()
        }
    }

}
